main_f4.c

```diff
@@ -205,10 +296,105 @@ board_init(void)
 
 	/* enable the power controller clock */
 	rcc_peripheral_enable_clock(&RCC_APB1ENR, RCC_APB1ENR_PWREN);
+}
+
+void
+board_deinit(void)
+{
+#if INTERFACE_USB
+	/* deinitialise GPIO9 (used to sniff VBUS) */
+	gpio_mode_setup(GPIOA, GPIO_MODE_INPUT, GPIO_PUPD_NONE, GPIO9);		
+#endif
 
+#if INTERFACE_USART
+	  /* deinitialise GPIO pins for USART transmit. */
+	  gpio_mode_setup(BOARD_PORT_USART, GPIO_MODE_INPUT, GPIO_PUPD_NONE, BOARD_PIN_TX | BOARD_PIN_RX);
+
+	  /* disable USART peripheral clock */
+	  rcc_peripheral_disable_clock(&BOARD_USART_CLOCK_REGISTER, BOARD_USART_CLOCK_BIT);
+#endif
+
+#if defined(BOARD_FORCE_BL_PIN_IN) && defined(BOARD_FORCE_BL_PIN_OUT)
+	/* deinitialise the force BL pins */
+	gpio_mode_setup(BOARD_FORCE_BL_PORT, GPIO_MODE_INPUT, GPIO_PUPD_NONE, BOARD_FORCE_BL_PIN_OUT);
+	gpio_mode_setup(BOARD_FORCE_BL_PORT, GPIO_MODE_INPUT, GPIO_PUPD_NONE, BOARD_FORCE_BL_PIN_IN);
+#endif
+
+#if defined(BOARD_FORCE_BL_PIN)
+	/* deinitialise the force BL pin */
+	gpio_mode_setup(BOARD_FORCE_BL_PORT, GPIO_MODE_INPUT, GPIO_PUPD_NONE, BOARD_FORCE_BL_PIN);
+#endif
+
+	/* deinitialise LEDs */	
+	gpio_mode_setup(
+		BOARD_PORT_LEDS, 
+		GPIO_MODE_INPUT, 
+		GPIO_PUPD_NONE,
+		BOARD_PIN_LED_BOOTLOADER | BOARD_PIN_LED_ACTIVITY);
+	
+	/* disable the power controller clock */
+	rcc_peripheral_disable_clock(&RCC_APB1ENR, RCC_APB1ENR_PWREN);
+
+	/* disable the GPIO port peripheral clocks */
+	rcc_peripheral_disable_clock(&RCC_AHB1ENR, RCC_AHB1ENR_IOPAEN);
+	rcc_peripheral_disable_clock(&RCC_AHB1ENR, RCC_AHB1ENR_IOPBEN);
+	rcc_peripheral_disable_clock(&RCC_AHB1ENR, RCC_AHB1ENR_IOPCEN);
+	rcc_peripheral_disable_clock(&RCC_AHB1ENR, RCC_AHB1ENR_IOPDEN);
+	rcc_peripheral_disable_clock(&RCC_AHB1ENR, RCC_AHB1ENR_IOPEEN);
+	rcc_peripheral_disable_clock(&RCC_AHB1ENR, RCC_AHB1ENR_IOPEEN);
+	rcc_peripheral_disable_clock(&RCC_AHB1ENR, RCC_AHB1ENR_IOPFEN);
+	rcc_peripheral_disable_clock(&RCC_AHB1ENR, RCC_AHB1ENR_IOPGEN);
+	rcc_peripheral_disable_clock(&RCC_AHB1ENR, RCC_AHB1ENR_IOPHEN);
+	rcc_peripheral_disable_clock(&RCC_AHB1ENR, RCC_AHB1ENR_IOPIEN);
```

Comment:
Should I individually reset each of the RCC peripheral clocks like what I am doing here, or should I just reset the entire AHB1, AHB2, APB1 and APB2 enable registers back to their defaults?